verus/lcf/src/verusport.rs

```diff
@@ -0,0 +1,1119 @@
+        }
+    }
+}
+
+impl SpecTerm {
+    //spec function returns true if the map contains the key when it a Var variant, or if it is Const variant
+    pub open spec fn spec_complete_subst(self, s: SpecSubst) -> bool {
+        match self {
+            SpecTerm::Var(v) => s.contains_key(v),
+            SpecTerm::Const(_) => true,
+        }
+    }
+
+    // spec function checks if SpecTerm is a base type
+    pub open spec fn spec_concrete(self) -> bool {
+        self is Const
+    }
+
+    //spec function reasons about substitution from Var variants to Const or does nothing if it is a Const variant
+    pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecTerm)
+        recommends
+            self.spec_complete_subst(s),
+    {
+        match self {
+            SpecTerm::Var(v) => {
+                let u_option = s.index(v);
+                SpecTerm::Const(u_option)
+            },
+            SpecTerm::Const(_) => self,
+        }
+    }
+}
+
+impl Term {
+    // function proves equality for Term types
+    pub fn term_eq(&self, other: &Self) -> (res: bool)
+        ensures
+            res <==> self.deep_view() == other.deep_view(),
+    {
+        match (self, other) {
+            (Term::Const(s), Term::Const(t)) => Const::const_eq(&s, &t),
+            (Term::Var(u), Term::Var(v)) => u.eq(v),
+            _ => false,
+        }
+    }
+
+    // function checks whether Term items are either Const or are Vars with a key in the Subst TmpStringHashMap
+    pub fn complete_subst(self, s: &Subst) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_complete_subst(s.deep_view()),
+    {
+        match self {
+            Term::Var(v) => s.contains_key(v.as_str()),
+            Term::Const(_) => true,
+        }
+    }
+
+    // function checks if term variant is a base type
+    pub fn concrete(self) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_concrete(),
+    {
+        match self {
+            Term::Const(_) => true,
+            Term::Var(_) => false,
+        }
+    }
+
+    // this functions performs a substitution to achieve a Const variant of Term
+    pub fn subst(self, s: &Subst) -> (res: Term)
+        requires
+            self.deep_view().spec_complete_subst(s.deep_view()),
+        ensures
+            res.deep_view().spec_concrete(),
+            res.deep_view() == self.deep_view().spec_subst(s.deep_view()),
+    {
+        match self {
+            Term::Var(v) => {
+                let u_option = s.get(v.as_str());
+                Term::Const(u_option.unwrap().clone())
+            },
+            Term::Const(_) => self,
+        }
+    }
+}
+
+//#[verifier::external_body] needed because the verifier does not support the equality relation we are trying to leverage
+#[verifier::external_body]
+//function checks whether 2 Vec<Term>s are equal or not
+pub fn terms_eq(a: &Vec<Term>, b: &Vec<Term>) -> (res: bool)
```

Comment:
It is possible to write a verified version of this function without axiomatizing---I would note this as a TODO item, you can assign it to me by adding a comment like `// TODO(pratap): verified version of this function`